Plugins/NativeCamera/Source/MediaStream.cpp

```diff
@@ -0,0 +1,335 @@
+#include "MediaStream.h"
+#include "NativeCameraImpl.h"
+#include <assert.h>
+
+namespace Babylon::Plugins
+{
+    arcana::task<Napi::Object, std::exception_ptr> MediaStream::New(Napi::Env env, Napi::Object constraints)
+    {
+        auto ctor = DefineClass(
+            env,
+            JS_CLASS_NAME,
+            {
+                InstanceMethod("getTracks", &MediaStream::GetVideoTracks), // The only supported tracks are video tracks
+                InstanceMethod("getVideoTracks", &MediaStream::GetVideoTracks),
+                InstanceMethod("getAudioTracks", &MediaStream::GetAudioTracks),
+                InstanceMethod("applyConstraints", &MediaStream::ApplyConstraints),
+                InstanceMethod("getCapabilities", &MediaStream::GetCapabilities),
+                InstanceMethod("getSettings", &MediaStream::GetSettings),
+                InstanceMethod("getConstraints", &MediaStream::GetConstraints),
+            });
+        
+        // Create a an object reference to the mediaStream javascript object so that it is not destructed during the async operation
+        auto mediaStreamObject{ std::make_shared<Napi::ObjectReference>(Napi::Persistent(ctor.New({}))) };
+        auto mediaStream{ MediaStream::Unwrap(mediaStreamObject->Value()) };
+        
+        return mediaStream->ApplyInitialConstraints(env, constraints).then(mediaStream->m_runtimeScheduler, arcana::cancellation::none(), [mediaStreamObject](const arcana::expected<void, std::exception_ptr>& result)
+        {
+            if (result.has_error())
+            {
+                throw result.error();
+            }
+
+            return mediaStreamObject->Value();
+        });
+    }
+
+    MediaStream::MediaStream(const Napi::CallbackInfo& info)
+        : Napi::ObjectWrap<MediaStream>{info}
+        , m_cameraImpl{std::make_shared<Camera::Impl>(info.Env(), false /* should finish removing overrideTexture */)}
+        , m_runtimeScheduler{JsRuntime::GetFromJavaScript(info.Env())}
+    {
+    }
+
+    Napi::Value MediaStream::GetVideoTracks(const Napi::CallbackInfo& info)
+    {
+        // To simply the implementation (because only a single track is supported by the native implementation) the same
+        // ObjectWrap fullfills the MediaStream and MediaStreamTrack interfaces
+        auto tracks = Napi::Array::New(info.Env(), 1);
+        tracks.Set(uint32_t{0}, info.This());
+        return std::move(tracks);
+    }
+
+    Napi::Value MediaStream::GetAudioTracks(const Napi::CallbackInfo& info)
+    {
+        // Only video tracks are currently supported
+        return Napi::Array::New(info.Env(), 0);
```

Comment:
Wrap in std::move?

Reply:
I don't think that's needed here because I'm directly returning the new object without storing it as a local variable first.